Tests/UnitTests/Core/Propagator/CovarianceEngineTests.cpp

```diff
@@ -164,5 +171,186 @@ BOOST_AUTO_TEST_CASE(covariance_engine_test) {
   BOOST_CHECK_NE(*(std::get<0>(boundResult).covariance()),
                  Covariance::Identity());
 }
+
+BOOST_AUTO_TEST_CASE(CovarianceConversionTest) {
+  GeometryContext gctx;
+
+  std::size_t n = 100;
+
+  std::mt19937 rng{42};
+
+  std::uniform_real_distribution<double> locDist(-5., 5.);
+  std::uniform_real_distribution<double> globDist(-50., 50.);
+  std::uniform_real_distribution<double> angleDist(-2 * M_PI, 2 * M_PI);
+  std::uniform_real_distribution<double> unitDist(0, 1);
+
+  for (std::size_t j = 0; j < 1; j++) {
+    Transform3 transformA =
+        Transform3::Identity() * AngleAxis3(M_PI_2, Vector3::UnitY());
+
+    transformA.translation() << globDist(rng), globDist(rng), globDist(rng);
+
+    auto planeSurfaceA = Surface::makeShared<PlaneSurface>(transformA);
+
+    auto planeSurfaceB = Surface::makeShared<PlaneSurface>(transformA);
+
+    auto boundToBound =
+        [&gctx](const auto& parIn, const auto& covIn, const auto& srfA,
+                const auto& srfB) -> std::pair<BoundVector, BoundMatrix> {
+      Acts::FreeVector freePars =
+          Acts::detail::transformBoundToFreeParameters(srfA, gctx, parIn);
+      Acts::BoundVector parOut =
+          Acts::detail::transformFreeToBoundParameters(freePars, srfB, gctx)
+              .value();
+
+      Acts::BoundToFreeMatrix boundToFreeJacobian =
+          srfA.boundToFreeJacobian(gctx, parIn);
+      Acts::FreeMatrix freeTransportJacobian = FreeMatrix::Identity();
+
+      FreeVector freeToPathDerivatives = FreeVector::Zero();
+      freeToPathDerivatives.head<3>() = freePars.segment<3>(eFreeDir0);
```

Comment:
Can you put Cross(dir, any_vector) for 4th - 6th elements and see the test is OK?

If I am correct, the test should pass.

Reply:
What's `any_vector` here?

Reply:
anything that is neither zero nor parallel to the direction, so that Cross(dir, anyvector) is not zero vector

Reply:
We can start with ones with orthogonal to the directino though

Reply:
If I add

```cpp
freeToPathDerivatives.segment<3>(4) = freePars.segment<3>(eFreeDir0).cross(Vector3::Random());
```

I get failures.

Reply:
For which cases and for which element? How big is the difference?

Reply:
You might need to put the same vector for both directions

Reply:
It is important because the `freeToPathDerivative` is (t_x,t_y,_tz,0, (txb)_x, (txb)_y, (txb)_z, 0) for both perigee and bound surface in ACTS implementation

Reply:
So that would mean the roundtrip works because we assume B is zero, but it's not clear if the actual covariance in the target plane surface or perigee is correct. 

Is there a good way to assert the target correctness? Other than comparing with an analytically calculated covariance?

Reply:
The way to assert the target correctness is comparing with numerical differentiation.

We can numerically obtain "J_{G->L} * Correction * J_T * J_{L->G}" which is the full jacobian from initial surface to target surface. (In your setup J_T is identity matrix.)

It is simply $$\frac{\partial x^{final}_j}{\partial x^{initial}_i} = \frac{x^{final}_j(x^{initial}_i +h) - x^{final}_j(x^{initial}_i)}{h}$$

Setting h~0.0001 works OK in my experience

I have been working on this in detray library - you can check here:
https://github.com/acts-project/detray/pull/610/files
Note that it is for the target correctness with two displaced surfaces.

---------------------------------------------------------------------------


Back to the main discussion,


> So that would mean the roundtrip works because we assume B is zero, 

This is not what I meant. 
we must get the consistent result even with non-zero bfield components, which are going to be cancelled anyway.
If you are not getting the consistent result with `cross(direction, bfield)` term, there could be a bug in the test.

The test of this PR has been OK without bfield because the free space transport Jacobian is just a Identity matrix. 
But if you are trying to do the same round-trip test with two "displaced" surfaces, not intersecting, it's not going to work anymore.

I can prove why ACTS implementation must be able to give you the same bound covariance after round trip for two "intersecting" surface, with `cross(direction,bfield)` term:
https://github.com/beomki-yeo/WhyRoundTripJacobianIsIdentityMatrix/releases/download/v1.0/main.pdf

Reply:
@beomki-yeo Ok, I added a fixed but random B field vector, and indeed that seems to succeed in all cases.

Reply:
OK just make sure that you don't hit the divergence conditions with random variables 😃

Reply:
Hmm maybe my derivations are wrong where I assumed J_{L -> G} J_{G -> L} is identity matrix at the final surface, which is not true. There must be more magic in the calculations :p ...

Reply:
Ok @beomki-yeo, I've done this now, but here it gets weird:

- With the direct jacobian from A->B I get the same covariance as when going trough J_{L->G} J_{G->L} **without the free-to-path derivatives!**
- With the direct jacobian, I **cannot roundtrip** A->B->A, it just gives different covariances back at A.

What's going on here?

Reply:
It's very interesting. I just have a guess here. 

It might be easy to explain with an image:

<img src="https://github.com/acts-project/acts/assets/63090140/6ff06d0f-6d4e-4888-b203-289491c2d0ec"  width="300" height="200" />

In the first case, I think you are calculating direct jacobians by projecting the displaced parameters onto the final surface vertically (A point in the image). This won't need a `free-to-path-derivative` 

If you want to include the `free-to-path-derivative` in jacobian calculation, you might need to project the parameter with the track direction ($\mathbf{t}$) (B point in the image). 

The second case can be explained in the same way. I believe you are getting A -> A' point during the direct jacobian calculation. This won't give you the identity matrix:
$$\frac{u(u_0+h) - u(u_0)}{h} \neq 1 $$

<img src="https://github.com/acts-project/acts/assets/63090140/a87c46f8-ad4a-456e-8c85-0104ee1d316d"  width="300" height="200" />




In summary, I believe the `localTolocal` in direction jacobian function is not being used correctly :thinking:

Reply:
Ok, so with the projection I get somewhat matching covariances. 

I think I'm observing some numerical instability: the top left 2x2 corners match at 10^-5, but the rest of the matrix, including the rest of the diagonal are only up to within 0.1 of the nominal calculation.

I think for the purpose of this test this is good enough. The main goal is to verify we can roundtrip covariances, with the direct jacobian being more of a sanity check.

@beomki-yeo @andiwand

Reply:
Ok for the last point, i suggest setting the cross(t×b) terms to zero because we are only projecting with t. (No propation under b field)

Reply:
Everything else looks good to me

Reply:
> Ok for the last point, i suggest setting the cross(t×b) terms to zero because we are only projecting with t. (No propation under b field)

I mean technically I could also set up an call a propagator, so that I get the correct BField projection from the propagation.

Reply:
Ok so with the numerical propagation, the resulting covariances are completely different w.r.t. to the nominal calculation.